File: src/webfrontend/src/models/Entry.ts
import { Comparable } from "./Comparable"
import { sameDate } from "../utils"

export class Entry implements Comparable<Entry>{
    public id: number = 0
    public title: string = ""
    public value: number = 0
    public date: Date = new Date()
    public categoryId: number | undefined = undefined

    constructor(id: number, 
        date: Date, 
        title: string, 
        value: number,
        categoryId: number = 0) {
        this.id = id
        this.date = date
        this.title = title
        this.value = value
        this.categoryId = categoryId
    }

    public Equals(other: Entry): boolean {
        return this.title === other.title
                && this.value === other.value
                && sameDate(this.date, other.date)
                && this.categoryId === other.categoryId
    }
}

File: src/Backend.Core/Models/Entry.cs
namespace Backend.Models
{
    public class Entry: EntryPlain
    {
        public Entry(): base()
        {
        }
        
        public Entry(EntryPlain other): base(other) { }

        public Entry(Entry other): base(other)
        {
            User = other.User;
        }
        public User User { get; set; }
        
        public Category Category { get; set; }
    }
    
    public class EntryPlain
    {
        public EntryPlain() { }
        public EntryPlain(EntryPlain other)
        {
            Value = other.Value;
            Date = other.Date;
            Title = other.Title;
            CategoryId = other.CategoryId;
            Id = other.Id;
        }

        public override bool Equals(object? obj)
        {
            if (obj is not EntryPlain other)
            {
                return false;
            }

            return Value == other.Value
                   && Date == other.Date
                   && Title == other.Title
                   && CategoryId == other.CategoryId
                   && Id == other.Id;
        }

        public int Id { get; set; }
        public float Value { get; set; }
        public DateTime Date { get; set; }
        public string Title { get;set; }
        public int UserId { get; set; }
        
        public int? CategoryId { get; set; }
    }
}
